Tests/Marketplace/search_and_install_packs.py

```diff
@@ -531,268 +721,282 @@ def search_pack_and_its_dependencies(client: demisto_client,
-    batch_packs_install_request_body: list = []  # List of lists of packs to install if not using multithreading .
-    # Each list contain one pack and its dependencies.
-    collected_dependencies: list = []  # List of packs that are already in the list to install.
-    master_commit_hash = get_env_var("LAST_UPLOAD_COMMIT")
-
-    lock = Lock()
-
-    kwargs = {
-        'client': client,
-        'packs_to_install': packs_to_install,
-        'installation_request_body': installation_request_body,
-        'lock': lock,
-        'collected_dependencies': collected_dependencies,
-        'production_bucket': production_bucket,
-        'multithreading': multithreading,
-        'batch_packs_install_request_body': batch_packs_install_request_body,
-        'commit_hash': master_commit_hash,
-    }
-
-    if not multithreading:
-        for pack_id in pack_ids:
-            search_pack_and_its_dependencies(pack_id=pack_id, **kwargs)
-
+    commit_hash: str = get_env_var("LAST_UPLOAD_COMMIT")
+
+    logging.info(f"Starting to search and install packs on server {host}")
+    start_time = datetime.utcnow()
+    success = True
+    all_packs_dependencies = {}
+    with ThreadPoolExecutor(max_workers=MAX_WORKERS if multithreading else 1) as pool:
+        futures = [
+            pool.submit(
+                get_pack_and_its_dependencies,
+                pack_id=pack_id,
+                client=client,
+                production_bucket=production_bucket,
+                commit_hash=commit_hash,
+            )
+            for pack_id in pack_ids
+        ]
+
+        for future in as_completed(futures):
+            try:
+                pack_id, dependencies_for_packs = future.result()
+                if dependencies_for_packs is not None:
+                    all_packs_dependencies[pack_id] = dependencies_for_packs
+                else:
+                    logging.error(
+                        f"Failed to search for dependencies of pack '{pack_id}'"
+                    )
+                    success = False
+            except Exception:  # noqa E722
+                logging.error(
+                    f"Failed to search for dependencies of pack '{pack_ids[futures.index(future)]}'"
+                )
+                success = False
+
+    if not success:
+        logging.critical(
+            "Failure while searching for packs dependencies, installing packs regardless."
+        )
+
+    logging.info(f"Starting to install packs on {host}")
+    # Gather all dependencies and install them in batches.
+    packs_installed_successfully: set[str] = set()
+    packs_to_install = []
+    for i, (pack_id, pack_dependencies) in enumerate(all_packs_dependencies.items()):
+        packs = flatten_dependencies(pack_id, pack_dependencies, all_packs_dependencies)
+        packs_to_install.extend(
+            [pack for pack in packs if pack["id"] not in packs_installed_successfully]
```

Comment:
If a pack failed to installed, it will not be in `packs_installed_successfully`, and thus you'll try to install it over and over again?
Maybe add it to a set called `packs_failed_install`, and check that as well?